python/ql/src/semmle/python/dataflow/new/internal/LocalSources.qll

```diff
@@ -33,15 +33,15 @@ private import DataFlowPrivate
 class LocalSourceNode extends Node {
   cached
   LocalSourceNode() {
-    not simpleLocalFlowStep(_, this) and
-    // Currently, we create synthetic post-update nodes for
-    // - arguments to calls that may modify said argument
-    // - direct reads a writes of object attributes
-    // Both of these preserve the identity of the underlying pointer, and hence we exclude these as
-    // local source nodes.
-    // We do, however, allow the post-update nodes that arise from object creation (which are non-synthetic).
-    not this instanceof SyntheticPostUpdateNode
+    this instanceof ExprNode and
+    not simpleLocalFlowStep(_, this)
     or
+    // Module variable nodes must be local source nodes, otherwise type trackers cannot step through
+    // them.
+    this instanceof ModuleVariableNode
+    or
+    // We explicitly include any read of a global variable, as some of these may have local flow going
+    // into them.
     this = any(ModuleVariableNode mvn).getARead()
```

Comment:
Thanks for adding these comments, makes it much easier :+1: 

Are there any instances where having `ModuleVariableNode` provides any value instead of _just_ having `any(ModuleVariableNode mvn).getARead()`? That is, could we write this part as:

```suggestion
    // Module variable nodes must be local source nodes, otherwise type trackers cannot
    // step through them. We explicitly include any read of a global variable, as some
    // of these may have local flow going into them.
    this = any(ModuleVariableNode mvn).getARead()
```

I tried running the following query on a few databases I had lying around (salt, nova, peewee), but it did not give any results.

```codeql
import python
import semmle.python.dataflow.new.internal.LocalSources
import semmle.python.dataflow.new.internal.DataFlowPublic
private import semmle.python.dataflow.new.internal.DataFlowPrivate

from ModuleVariableNode mvn, Node flowsToNode
where
  not mvn = flowsToNode and
  mvn.(LocalSourceNode).flowsTo(flowsToNode) and
  not mvn.getARead().(LocalSourceNode).flowsTo(flowsToNode)
select mvn, flowsToNode
```

I guess I could just try to run our tests with that change, so I've just started doing that :stuck_out_tongue:

Reply:
I just tried out running the tests, and some of them failed. For example https://github.com/github/codeql/blob/6823855e9cac61a72b934a2b8b1f99b246d656f1/python/ql/test/experimental/dataflow/ApiGraphs/test.py#L50

Reply:
The trouble arises when you want to type track from `some_global = tracked` to `sink(some_global)`. In practice this happens in two steps: first we go from `tracked` to `ModuleVariableNode for some_global` and then from this node to the read.

However, since type trackers (in their usual formulation) step from `LocalSourceNode` to `LocalSourceNode`, we'll lose the first step unless the `ModuleVariableNode` is also a `LocalSourceNode`.

On the other hand, we _should_ eventually be able to get rid of the `getARead` bit. It's only there right now because we have local flow into global variables, which I think we should get rid of (but we _do_ want it in the global scope, so this is slightly tricky).

For instance we currently have local flow from source to sink in the following bit of code:

```python
def make_safe():
    global g
    g = NON_SOURCE

def foo():
    global g
    g = SOURCE
    make_safe()
    sink(g)
```
Now, we'll actually have this regardless (unless we put some effort into tracking what globals are modified by what functions, which we might), since there will be the aforementioned step through the module variable node. But we _shouldn't_ have local flow here.

I think based on this discussion that I should maybe expand the comment for `ModuleVariableNode` a bit. :sweat_smile:

Reply:
#6218 (which is based on this branch) clarifies this situation a bit by creating a specific `TypeTrackingNode` class that encompasses both `LocalSourceNode` and `ModuleVariableNode`. That way, we don't have the somewhat confusing inclusion of `ModuleVariableNode` as a `LocalSourceNode`.

Reply:
Should this PR be closed then, and we focus on the other one instead?

Reply:
I think it makes more sense to merge this one as is (assuming you're happy with it), as the changes it makes have been tested (performance and otherwise), whereas the performance tests for the other PR are still ongoing. Also, the other changes will affect the shared type tracking library, so are perhaps a bit more controversial.

Reply:
alright, fine by me :+1: